dash-renderer/src/persistence.js

```diff
@@ -0,0 +1,497 @@
+ *
+ * If any `persisted_props` are nested prop IDs, the component should define a
+ * class property (not a React prop) `persistenceTransforms`, as an object:
+ * {
+ *   [propName]: {
+ *     [piece]: {
+ *       extract: propValue => valueToStore,
+ *       apply: (storedValue, propValue) => newPropValue
+ *     }
+ *   }
+ * }
+ * - `extract` turns a prop value into a reduced value to store.
+ * - `apply` puts an extracted value back into the prop. Make sure this creates
+ *   a new object rather than mutating `proValue`, and that if there are
+ *   multiple `piece` entries for one `propName`, their `apply` functions
+ *   commute - which should not be an issue if they extract and apply
+ *   non-intersecting parts of the full prop.
+ * You only need to define these for the props that need them.
+ * It's important that `extract` pulls out *only* the relevant pieces of the
+ * prop, because persistence is only maintained if the extracted value of the
+ * prop before applying persistence is the same as it was before the user's
+ * changes.
+ */
+
+import {
+    equals,
+    filter,
+    forEach,
+    keys,
+    lensPath,
+    set,
+    symmetricDifference,
+    type,
+} from 'ramda';
+import {createAction} from 'redux-actions';
+import uniqid from 'uniqid';
+
+import Registry from './registry';
+
+const storePrefix = '_dash_persistence.';
+const UNDEFINED = 'U';
+
+function err(e) {
+    const error = typeof e === 'string' ? new Error(e) : e;
+
+    /* eslint-disable no-console */
+    // Send this to the console too, so it's still available with debug off
+    console.error(e);
+    /* eslint-disable no-console */
+
+    return createAction('ON_ERROR')({
+        myUID: uniqid(),
+        myID: storePrefix,
+        type: 'frontEnd',
+        error,
+    });
+}
+
+/*
+ * Does a key fit this prefix? Must either be an exact match
+ * or, if a separator is provided, a scoped match - exact prefix
+ * followed by the separator (then anything else)
+ */
+function keyPrefixMatch(prefix, separator) {
+    const fullStr = prefix + separator;
+    const fullLen = fullStr.length;
+    return key => key === prefix || key.substr(0, fullLen) === fullStr;
+}
+
+class WebStore {
+    constructor(backEnd) {
+        this._name = backEnd;
+        this._storage = window[backEnd];
+    }
+
+    hasItem(key) {
+        return this._storage.getItem(storePrefix + key) !== null;
+    }
+
+    getItem(key) {
+        // note: _storage.getItem returns null on missing keys
+        // and JSON.parse(null) returns null as well
+        const gotVal = this._storage.getItem(storePrefix + key);
+        return gotVal === UNDEFINED ? void 0 : JSON.parse(gotVal);
+    }
+
+    /*
+     * In addition to the regular key->value to set, setItem takes
+     * dispatch as a parameter, so it can report OOM to devtools
+     */
+    setItem(key, value, dispatch) {
+        const setVal = value === void 0 ? UNDEFINED : JSON.stringify(value);
```

Comment:
First time I've seen someone use `void 0` in a while in JS :)

As ES5 explicitly states that `undefined` is read-only, I don't think we need to protect against reassignment.

IE9 and Edge10 have full ES5 support so I don't think this is a concern.

Was there an additional concern here?

Reply:
yeah, just satisfying the linter but you're right, it would be better to just disable that check in the linter.

Reply:
🔪 no-undefined https://github.com/plotly/dash/pull/903/commits/15ca0e5ae6747af2fb8e5c898a89c48f4ddef0a3